FEAT: aarch64, ppc64le and linux64 from source

MementoRC opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

MementoRC commented:
@Tobias-Fischer This will fail with the same error as for a `linux-x86_64`, which will hopefully be solved by the next update of `py-lief>0.14.1`

wolfv commented:
impressive work @MementoRC !

MementoRC commented:
I think this is pretty much ready. LIEF project says that the badAlloc issue is resolved:
https://github.com/lief-project/LIEF/issues/1060

MementoRC commented:
@Tobias-Fischer @wolfv This PR includes both `linux64` and `aarch64`. It uses the previous conda version of `sbcl` as bootstrap, then compiles itself.

Tobias-Fischer commented:
Hi @MementoRC  - can you please tidy up non used code, and add some comments on what is being done in the build.sh etc? Many thanks!

MementoRC commented:
Sure, I will also check the windows build, the redefinition are probably not needed

Tobias-Fischer commented on recipe/meta.yaml:
```diff
@@ -1,84 +1,90 @@
-      - patches/0002-win-update-flags-zstd.patch  # [win and x86_64]
+      - patches/0001-win-update-flags-zstd.patch  # [win and x86_64]
+
+      - patches/0001-cross-use-qemu.patch  # [aarch64 or ppc64le]
```
This should probably `build!=target`?

MementoRC replied:
cross-compiling for windows arm64 fails because "CROSSCOMPILING_EMULATOR" is not set, so the selector needs to be more selective.

Tobias-Fischer commented on recipe/patches/0001-linux64-update-flags-zstd.patch:
```diff
@@ -0,0 +1,19 @@
+ endif
+
+-CFLAGS += -Wunused-parameter -fno-omit-frame-pointer -momit-leaf-frame-pointer
++CFLAGS += -Wunused-parameter -fno-omit-frame-pointer -momit-leaf-frame-pointer
```
This does nothing, right? In that case, the zstd-flags patches can be combined between linux-64 and linux-aarch64?

MementoRC replied:
Correct, it's an oversight from the debugging. Yes, we could combine a few patches, though it may render the selector a bit long. I combined the x86_64 and aarch64 zstd, but kept the osx and win separate. The cross selector should not be applied when cross-compiling on windows
I removed the `ppc64le`, I have no clue as to why it does not work after working on it for quite a bit, I am hoping something will come from the SBCL community, so likely it'll be a while before it can be integrated

Tobias-Fischer reviewed (commented):
Just a few minor things.

MementoRC commented:
@Tobias-Fischer @wolfv I have added running the full testsuite. However, a few tests fail on linux, more fails on win and emulated tests on aarch64 take too long. It also makes the recipe quite complicated

Also, I added the PPC64LE build that worked on Travis (I wasn't able to understand how to make it cross-compile with QEMU)

MementoRC commented:
`aarch64` builds on Travis and can run the test suite. I feel it provides a more valuable build (with info on which tests fail) than the cross-build. Thoughts?

MementoRC commented:
@Tobias-Fischer Any concerns/changes for this PR?

wolfv commented:
Is this one good to go?! :)

Tobias-Fischer commented:
I’m on leave and won’t have time to look into it for another little while - if you could please take a look and merge @wolfv that’d be great

MementoRC commented:
Cool, enjoy your leave, I eventually guessed that was the case. Yes, it's good enough to go (before I start adding new stuff!) - The only oddity is that the makefile for windows has an odd issue in interpreting `-lzstd` as a target when building the dynamic library.
This feature is not in the upstream, anyway. Thanks

wolfv commented on recipe/patches/0002-linux-patchelf.patch:
```diff
@@ -0,0 +1,10 @@
+     echo Unknown option \'"$1"\' to make-target-2
+     exit 1
+ fi
++patchelf --add-rpath $PREFIX/lib ./src/runtime/sbcl
```
I am wondering a bit wether this could be done in the build script, after everything is done?

MementoRC replied:
If I remember correctly, because `sbcl` first bootstrap itself, then use that bootstrap to build itself, the build would fail in the middle.

wolfv commented:
🎉 excellent work!
